### ISSUE TITLE
_compute_task_keep_ratio silently saturates to 1.0, breaking mask isolation past capacity

### ISSUE THREAD

**Issue by sachncs** [sachncs/tsn-affinity#6]
## Problem

`TSNBaseStrategy._compute_task_keep_ratio` returns `base / free_ratio` clamped only to `[1e-3, 1.0]` for the upper bound. When the occupied weight ratio grows past `1 - base`, `free_ratio` becomes smaller than `base`, the ratio exceeds 1.0, and is silently clamped to 1.0, leaving the strategy with no sparsity enforcement.

## Evidence

`tsn_affinity/strategies/tsn_base.py:192-213`:

```python
def _compute_task_keep_ratio(self) -> float:
    base = float(self.sparse_config.keep_ratio)
    ...
    if total == 0:
        return float(max(1e-3, min(1.0, base)))
    ...
    task_keep_ratio = base / free_ratio
    return float(max(1e-3, min(1.0, task_keep_ratio)))
```

Example: with `keep_ratio=0.3` and three consecutive tasks each occupying 25% of the mask, `free_ratio=0.25`, `task_keep_ratio=0.3/0.25=1.2`, clamped to 1.0. The new task keeps 100% of the unoccupied weights, defeating the very isolation the mask is supposed to provide.

`tsn_affinity/sparse/topk.py:42-44` handles `keep_ratio >= n_free` by returning `flat_free` unchanged (i.e. all remaining slots kept), so the downstream TopK STE does not raise — it silently degrades to dense training. The masking contract documented at `tsn_affinity/sparse/linear.py:13-25` is silently violated.

## Root Cause

The intent of the formula is "redistribute the budget so the union remains `base`". But the implementation assumes `free_ratio >= base` for every additional task. After enough tasks this assumption breaks, and the clamp to 1.0 hides the failure.

## Impact

- For `n_tasks > ceil(1 / keep_ratio)`, the mask union saturates and `keep_ratio` becomes a no-op.
- BWT (Backward Transfer) and forgetting metrics reported by `tsn_affinity/run/analysis.py` on saturated masks no longer measure continual learning — they measure plain sequential fine-tuning with random initialisation per task.
- Users running synthetic benchmarks with `n_tasks=10` and `keep_ratio=0.3` (as documented in `README.md:139`) silently hit saturation at task 4.

## Recommendation

Raise `RuntimeError` (or a dedicated `MaskError` from `tsn_affinity/core/exceptions.py`) when the requested keep ratio would exceed the available capacity; alternatively, redistribute `base` evenly across `n_tasks` at strategy initialisation so the budget is honest from the start. The clamp-to-1.0 silent-degrade path should not survive.

## Acceptance Criteria

- [ ] When `_compute_task_keep_ratio` would exceed 1.0, the strategy either (a) raises a clear error, or (b) recomputes `keep_ratio` to `free_ratio` and emits a warning
- [ ] A test sets `keep_ratio=0.3`, runs 10 tasks, and asserts the union never exceeds the initial mask budget

## Verification

`pytest -k "keep_ratio" -v` should pass with the new test. Re-run `tsn-benchmark --n-tasks 10 --keep-ratio 0.3` and inspect `performance_matrix.npy` to confirm mask union stays within budget.

## Risk

Medium. Changing silent-degrade to a hard error can break benchmarks that previously ran; reviewers should compare BWT/Forgetting numbers on the existing `runs/benchmark_2026_04_30/` baseline.

**Comment by sachncs** [comment]
Fixed in commit a799277. _compute_task_keep_ratio now raises ConfigurationError when mask capacity is exhausted (free_ratio <= 1e-6) or when the effective keep_ratio would saturate (>= 1.0 - 1e-6). _prepare_current_task wraps the error as StrategyError so multi-task callers see a uniform exception type. The new test_capacity_exhaustion_raises test verifies the failure path.
